daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala

```diff
@@ -1290,8 +1302,15 @@ private[lf] object SBuiltin {
           // Check if we have a cached global key result.
           onLedger.ptx.globalKeyInputs.get(gkey) match {
             case Some(optCid) =>
-              onLedger.ptx = onLedger.ptx.copy(keys = onLedger.ptx.keys.updated(gkey, optCid))
-              optCid match {
+              val optActiveCid = optCid.flatMap { cid =>
+                if (onLedger.ptx.consumedBy.contains(cid)) {
+                  KeyInactive
+                } else {
+                  KeyActive(cid)
+                }
+              }
```

Comment:
Not a big fan of adding a `flatMap` method only for that.

Reply:
6d05e4c41660720c4ce494d3f057612de9978261